### PR TITLE
Vocabbuilder.koplugin: add search ability

### PR DESCRIPTION
**Pull request by weijiuqiao**
This PR adds the ability to search words from the vocab builder plugin.

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/9881)

### CODE REVIEW

File: plugins/vocabbuilder.koplugin/db.lua [hunk 1]
```diff
@@ -31,6 +31,15 @@ local VOCABULARY_DB_SCHEMA = [[
     CREATE INDEX IF NOT EXISTS title_name_index ON title(name);
 ]]
 
+local function convertToSqlLike(str)
```
[the remark this answers is not in the document]
Comment (weijiuqiao) [reply plugins/vocabbuilder.koplugin/db.lua:34]: It's not just escape, if no wildcard is used, the str will be enclosed by two %'s by default, thus `convert`ing.

[the remark this answers is not in the document]
Comment (Frenzie) [reply plugins/vocabbuilder.koplugin/db.lua:34]: Prepare maybe, or is that too loaded in an SQL context? Convert doesn't really jive with me here. Anyway, let's ask @NiLuJe.

[the remark this answers is not in the document]
Comment (NiLuJe) [reply plugins/vocabbuilder.koplugin/db.lua:34]: `massageForSQL`?

[the remark this answers is not in the document]
Comment (NiLuJe) [reply plugins/vocabbuilder.koplugin/db.lua:34]: Being somewhat SQL-illiterate, I'm not quite sure what the % quoting does, but `escape` sounds good?

[the remark this answers is not in the document]
Comment (NiLuJe) [reply plugins/vocabbuilder.koplugin/db.lua:34]: Or something based on `sanitize`, although that may not entirely the goal here?

[the remark this answers is not in the document]
Comment (Frenzie) [reply plugins/vocabbuilder.koplugin/db.lua:34]: % is a wildcard so it's a bit like `.*we want this.*`

File: plugins/vocabbuilder.koplugin/main.lua [hunk 2]
```diff
@@ -1700,10 +1799,9 @@ function VocabularyBuilderWidget:onSwipe(arg, ges_ev)
 end
 
 function VocabularyBuilderWidget:onMultiSwipe(arg, ges_ev)
-    -- For consistency with other fullscreen widgets where swipe south can't be
-    -- used to close and where we then allow any multiswipe to close, allow any
-    -- multiswipe to close this widget too.
-    self:onClose()
+    -- since swipe south is closing, we reload items here
+    self:reloadItems()
```
Comment (poire-z) [inline plugins/vocabbuilder.koplugin/main.lua:1803]: If I were a Vocabulary Builder user, I would be using multiswipe to quit it, and I would really complain loudly about the inconsistency :)
As I am not, I will just complain silently :)
(If it were me, and I dunno how useful reloading is, I would probably try to match all 4 of "north/west/south/east" in the gesture name, expecting the user to draw a square/circle in any direction/order, like a reload icon:
image )

File: plugins/vocabbuilder.koplugin/db.lua (hunk 1 above)
[the remark this answers is not in the document]
Comment (weijiuqiao) [reply plugins/vocabbuilder.koplugin/db.lua]: Naming is always so hard. So I removed this functions and moved the preparation to when the search text is set :) It also reduces the number of times to prepare it.

```
if self.search_text == "" then
    self.search_text_sql = nil
elseif self.search_text:find("%", 1, true) or self.search_text:find("_") then
    self.search_text_sql = self.search_text:gsub("'", "''")
else
    self.search_text_sql = "%" .. self.search_text:gsub("'", "''") .. "%"
end
```

File: plugins/vocabbuilder.koplugin/main.lua (hunk 2 above)
Comment (weijiuqiao) [reply plugins/vocabbuilder.koplugin/main.lua]: Good idea. Your silent complaint has been heard :) and the circle gesture done (though not sure if it is the preferred way to match directions)

Comment (poire-z) [reply plugins/vocabbuilder.koplugin/main.lua]: Thanks :)
My "restart koreader" gesture is south/west/north/east - so you could allow starting the circle with south. (Starting it with north is quite a bit less natural and harder for my brain to do :)

Comment (weijiuqiao) [reply plugins/vocabbuilder.koplugin/main.lua]: Done.

**Pull request merged**